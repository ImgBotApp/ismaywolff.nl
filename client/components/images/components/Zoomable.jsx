import React, { PropTypes } from 'react'
import ImageZoom from 'react-medium-image-zoom'
import { connect } from 'react-redux'
import { getImageEntities } from '../selectors'
import { createUrl, getAvailableWidth, roundUp } from '../utils'
import Placeholder from './Placeholder'

export function Zoomable({ entities, id, width }) {
  const image = entities[id]

  if (!image) {
    return <Placeholder height={1} width={1} />
  }

  return (
    <Placeholder height={1} width={image.width / image.height}>
      <ImageZoom
        image={{
          src: createUrl({ url: image.url, width }),
          alt: image.title
        }}
        zoomImage={{
          src: createUrl({
            url: image.url,
            width: roundUp(getAvailableWidth({ image, window }))
          }),
          alt: image.title
        }}
      />
    </Placeholder>
  )
}

Zoomable.propTypes = {
  entities: PropTypes.object.isRequired,
  id: PropTypes.string.isRequired,
  width: PropTypes.number.isRequired
}

const mapStateToProps = state => ({
  entities: getImageEntities(state)
})

export default connect(mapStateToProps)(Zoomable)
