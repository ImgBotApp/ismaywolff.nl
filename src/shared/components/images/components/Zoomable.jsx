/* global window */

import React from 'react'
import { string, number, objectOf, object } from 'prop-types'
import ImageZoom from 'react-medium-image-zoom'
import { connect } from 'react-redux'
import { getImageEntities } from '../selectors'
import { createUrl, getAvailableWidth, getImageWidth } from '../utils'
import Placeholder from './Placeholder'

/**
 * Fallback viewport size when window isn't available (so on the server).
 * Taken from: http://gs.statcounter.com/screen-resolution-stats
 */

const fallbackViewport = {
  height: 640,
  width: 360
}

export function DumbZoomable({ entities, id, width }) {
  const image = entities[id]
  /* istanbul ignore next: window isn't available when testing */
  const viewport = {
    height: typeof window === 'object' ? window.innerHeight : fallbackViewport.height,
    width: typeof window === 'object' ? window.innerWidth : fallbackViewport.width
  }

  if (!image) {
    return <Placeholder height={1} width={1} />
  }

  return (
    <Placeholder height={1} width={image.width / image.height}>
      <ImageZoom
        image={{
          src: createUrl({ url: image.url, width }),
          alt: image.title
        }}
        zoomImage={{
          src: createUrl({
            url: image.url,
            width: getImageWidth(getAvailableWidth({ image, viewport }))
          }),
          alt: image.title
        }}
        shouldPreload
      />
    </Placeholder>
  )
}

DumbZoomable.propTypes = {
  entities: objectOf(object).isRequired,
  id: string.isRequired,
  width: number.isRequired
}

const mapStateToProps = state => ({
  entities: getImageEntities(state)
})

export default connect(mapStateToProps)(DumbZoomable)
