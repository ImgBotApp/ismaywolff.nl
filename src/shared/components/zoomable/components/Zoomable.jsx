/* global window */

import React from 'react'
import { string, shape, number } from 'prop-types'
import ImageZoom from 'react-medium-image-zoom'
import { utils } from '../../../data/images'
import { Placeholder } from '../../placeholder'
import ZoomableGridItem from './ZoomableGridItem'

/**
 * Fallback viewport size when window isn't available (so on the server).
 * Taken from: http://gs.statcounter.com/screen-resolution-stats
 */

const fallbackViewport = {
  height: 640,
  width: 360
}

const Zoomable = ({ image }) => {
  /* istanbul ignore next: window isn't available when testing */
  const viewport = {
    height: typeof window === 'object' ? window.innerHeight : fallbackViewport.height,
    width: typeof window === 'object' ? window.innerWidth : fallbackViewport.width
  }

  /**
   * Fallback for browsers that don't support responsive images
   */

  const src = utils.createUrl({ url: image.url, width: 250 })

  /**
   * The available image sizes for the browser to choose from
   */

  const srcSet = `
    ${utils.createUrl({ url: image.url, width: 1500 })} 1500w,
    ${utils.createUrl({ url: image.url, width: 1250 })} 1250w,
    ${utils.createUrl({ url: image.url, width: 1000 })} 1000w,
    ${utils.createUrl({ url: image.url, width: 750 })} 750w,
    ${utils.createUrl({ url: image.url, width: 500 })} 500w,
    ${utils.createUrl({ url: image.url, width: 250 })} 250w
  `

  /**
   * The breakpoints that influence how wide the thumbnail images are displayed, from the bottom up
   * it's:
   *
   * - By default images are displayed at one image per row (so 100vw)
   * - From 25em wide the grid snaps to two images per row (so 50vw)
   * - From 30em wide the wrapping container constrains the width to 30rem & 2 img (so max is 15rem)
   * - From 40em wide the grid snaps to three images per row (so 10rem)
   */

  const sizes = `
    (min-width: 40em) 10rem,
    (min-width: 30em) 15rem,
    100vw
  `

  return (
    <ZoomableGridItem>
      <Placeholder height={1} width={image.width / image.height}>
        <ImageZoom
          image={{
            src,
            srcSet,
            sizes,
            alt: image.title
          }}
          zoomImage={{
            src,
            srcSet,
            sizes: `${utils.getRatio({ image, viewport })}vw`,
            alt: image.title
          }}
        />
      </Placeholder>
    </ZoomableGridItem>
  )
}

Zoomable.propTypes = {
  image: shape({
    title: string.isRequired,
    url: string.isRequired,
    width: number.isRequired,
    height: number.isRequired
  }).isRequired
}

export default Zoomable
