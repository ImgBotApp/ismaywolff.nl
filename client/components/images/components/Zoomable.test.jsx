import React from 'react'
import { shallow } from 'enzyme'
import { DumbZoomable } from './Zoomable'

describe('<Zoomable />', () => {
  it('renders correctly', () => {
    const entities = { one: { url: 'url', title: 'title' } }
    const wrapper = shallow(<DumbZoomable entities={entities} id={'one'} width={100} />)
    expect(wrapper).toMatchSnapshot()
  })

  it('renders a loading state', () => {
    const wrapper = shallow(<DumbZoomable entities={{}} id={'one'} width={100} />)
    expect(wrapper).toMatchSnapshot()
  })
})
