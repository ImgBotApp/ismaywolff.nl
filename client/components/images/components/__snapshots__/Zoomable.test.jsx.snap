// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Zoomable /> renders a loading state 1`] = `
<glamorous(picture)
  height={1}
  width={1}
/>
`;

exports[`<Zoomable /> renders correctly 1`] = `
<glamorous(picture)
  height={1}
  width={NaN}
>
  <ImageZoom
    defaultStyles={
      Object {
        "image": Object {},
        "overlay": Object {},
        "zoomContainer": Object {},
        "zoomImage": Object {},
      }
    }
    image={
      Object {
        "alt": "title",
        "src": "url?w=100&fl=progressive",
      }
    }
    isZoomed={false}
    onUnzoom={[Function]}
    onZoom={[Function]}
    shouldHandleZoom={[Function]}
    shouldPreload={true}
    shouldReplaceImage={true}
    shouldRespectMaxDimension={false}
    zoomImage={
      Object {
        "alt": "title",
        "src": "url?w=1050&fl=progressive",
      }
    }
    zoomMargin={40}
  />
</glamorous(picture)>
`;
